### DISCUSSION TITLE
Task - Spring Boot - Partial CRUD API  with Service Layer - my_org DB

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#16] [Tasks]
# Request Handling with n-Tier CRUD API Having a Service Layer

## Introducing Service Layer

Create a Spring Boot application named `springBootCrudApi` (artifactId and groupId are in all lowercase) to connect with the `org` DB using **Jdbc** (mysql-connector-j). Create the base package of the project named `com.springboot.workers.crudapi`.

The details for creating the Spring Boot project are as follows:
- name: `springBootCrudApi`
- group: `com.workers`
- artifact: `springbootcrudapi`
- package: `com.springboot.workers.crudapi`

Dependencies to include:
- Spring Web
- MySQL Driver
- Spring Boot DevTools (optional)
- Spring Configuration Processor (optional)

    ```xml
    <dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>

		<dependency>
			<groupId>mysql</groupId>
			<artifactId>mysql-connector-java</artifactId>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
    </dependencies>
    ```

Perform _CRUD_ operations on the `Worker` table using models, DAOs, and repositories. 

Create these sub-packages under the base package:
  - `com.springboot.workers.crudapi.util`
  - `com.springboot.workers.crudapi.model`
  - `com.springboot.workers.crudapi.dao`
  - `com.springboot.workers.crudapi.repository`
  -  `com.springboot.workers.crudapi.service`
  - `com.springboot.workers.crudapi.controller`

**Hint - You can refer to discussion #15  for code snippets to reuse.**

Create the helper classes such as `DatabaseConnection`, `DateTimeUtilities` in the `com.springboot.workers.crudapi.util` package`.

Create the following **model** classes in `com.springboot.workers.crudapi.controller` package:
- `Worker`

Create the following **dao** interfaces in `com.springboot.workers.crudapi.dao` package:
- `WorkerDAO`: In the dao, apart from the previous abstract methods, include `boolean updateWorkerEmail(String email)` and rename the `update` method present earlier to `replace`.

Create the following **repository** classes in `com.springboot.workers.crudapi.repository` package:
- `WorkerRepository`: Implement the abstract methods of `WorkerDAO` to perform the DB operations.

Create the following **service** classes in `com.springboot.workers.crudapi.service` package:
- `WorkerService`: In this Service class marked with `@Service` annotation, autowire the `WorkerRepository`, and use the methods of the `WorkerRepository` class to perform the respective operations.

Create the following **controller** classes in `com.springboot.workers.crudapi.controller` package:
- `HomeController`
- `WorkerController`: In this Controller class, autowire the `WorkerService`. The controller should be thin and just have minimum calls and statements.

Create the handler methods and the following mappings in the appropriate controllers:
- `/` - GET - Displays a welcome message "Hello visitor!\nVisiting time: ${CURRENT_TIMESTAMP}"
- `/worker/worker/:id` - **GET** - Displays the details of a `Worker` record. Use the `worker_id` from the path variable to display the worker details. Return the worker record as a JSON response.
- `/worker/all` - **GET** - Displays the details of all the workers in the `Worker` table. Return the List of the workers as a JSON response.
- `/worker/create` - **POST** - Create a new worker record. Use the data from RequestBody(raw JSON) to insert a new worker record. Return `true` if the worker was successfully created, else it returns `false`. Send a _201_ (HttpStatus.CREATED) response code upon successful insertion.
- `/worker/update/:id` - **PATCH** - Update the `email` of the worker with the same `worker_id` as the one in the path variable to "mfs.akash@gmail.com". Take the new email in the request body. Return `true` if the record was updated successfully, else return `false`.
- `/worker/delete/:id` - **DELETE** - Delete the record of the worker whose `worker_id` matches `:id` in the path variable. Return `true` if the record was successfully deleted, else return `false`.

Use Postman to test the API.

**Comment by RohitN09** [comment]
https://github.com/RohitN09/Spring/tree/master/springBootCrudApi

**Comment by preeti-max** [comment]
https://github.com/preeti-max/crudapi/tree/master/dataAccessWithJdbc

**Comment by ratnadeepikavuddagiri** [comment]
https://github.com/ratnadeepikavuddagiri/SpringProjects/tree/main/springBootCrudApi

**Comment by shreiyarandive** [comment]
https://github.com/shreiyarandive/spring-java/tree/main/springBootRequestHandling

**Comment by Gaurav-Joshi-31** [comment]
https://github.com/Gaurav-Joshi-31/springprojects/tree/main/springBootCrudApi

**Comment by Shubhi27** [comment]
https://github.com/Shubhi27/spring_projects/tree/main/springBootCrudApi

**Comment by Sarfinaa** [comment]
[Code](https://github.com/Sarfinaa/SpringAndSpringBootProjects/tree/master/springBootCrudApi)

**Comment by Mrafe** [comment]
<a href="https://github.com/Mrafe/Spring/tree/main/springBootCrudApi">Spring Boot CRUD API with Service added</a>

**Comment by rakeshGit419** [comment]
[https://github.com/rakeshGit419/Wiley_Java/tree/main/OneDrive/Desktop/Wiley_java/spring-boot-my-org-crud-api](url)

**Comment by Batchusatyasri** [comment]
[spring boot crud api](https://github.com/Batchusatyasri/spring_core/tree/main/springBootCrudApi)

**Comment by amruth2211** [comment]
https://github.com/amruth2211/Spring-Projects/tree/main/springBootCrudApi

**Comment by abhiraj2k** [comment]
https://github.com/abhiraj2k/spring-projects/tree/master/springbootcrudapi

**Comment by Ashwesh09** [comment]
[Added services - My submission ](https://github.com/Ashwesh09/Web-Projects/tree/main/springbootjdbctemplate)

**Comment by mkaifahmed** [comment]
https://github.com/mkaifahmed/springcrud-with-service

**Comment by PavanArvapally** [comment]
https://github.com/PavanArvapally/spring-projects/tree/main/springBootCrudApi

**Comment by SrideviYamala** [comment]
https://github.com/SrideviYamala/spring-projects/tree/master/springBootCrudApi

**Comment by Harshita-Joshi** [comment]
https://github.com/Harshita-Joshi/Spring/tree/main/springBootCrudApi

**Comment by KaminiJha** [comment]
https://github.com/KaminiJha/SpringBootTasks/tree/main/springBootCrudApi

**Comment by lalit-ky** [comment]
https://github.com/009lalitkumar/Spring-SpringBoot-Tasks/tree/main/16

**Comment by Khyati-tripathi** [comment]
https://github.com/Khyati-tripathi/Spring/tree/main/springBootCrudApi

**Comment by Prem-3** [comment]
[https://github.com/Prem-3/Task/tree/main/spring-boot-my-org-crud-api](url)

**Comment by ujjwal18bit1167** [comment]
https://github.com/ujjwal18bit1167/Spring-Framework/tree/main/springBootCrudApi
